### PR TITLE
Connection pool implementation for Postgres

### PR DESCRIPTION
**Pull request by suresh-prakash**
## Description
Implement a connection pool for Postgres for non-auto-commit connections.
Updated the configuration reading

### Testing
Added unit test for config. reading. All the unit and integration tests are passing after the connection pool implementation.

### Checklist:
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by kotharironak** [comment]
@saxenakshitiz whenever you get sometime can you check this one as you did some testing?

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java [hunk 1]
```diff
@@ -427,7 +432,7 @@ public Optional<Document> update(
       throw new IOException("At least one update is required");
     }
 
-    try (final Connection connection = client.getNewConnection()) {
+    try (final Connection connection = connectionPool.getConnection()) {
```
Comment (saxenakshitiz) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java:435]: Can we expose getPooledConnection() in PostgresClient and use it here? It abstract out the logic of connection management to PostgresClient, and PostgresCollection doesn't have to deal with it?

**Comment by saxenakshitiz** [comment]
I am fine with current approach. However it will be good to keep connection management in PostgresClient.

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java (hunk 1 above)
Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java:435]: Makes sense. Will update it.

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresDatastore.java [hunk 2]
```diff
@@ -26,47 +26,19 @@ public class PostgresDatastore implements Datastore {
...
-
-      String finalUrl = url + this.database;
-      client =
-          new PostgresClient(
-              finalUrl, user, password, maxConnectionAttempts, connectionRetryBackoff);
+      final PostgresConfig postgresConfig = ConfigBeanFactory.create(config, PostgresConfig.class);
 
+      client = new PostgresClient(postgresConfig);
+      database = postgresConfig.getDatabase();
+      connectionPool = new PostgresConnectionPool(postgresConfig);
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresDatastore.java:41]: We don't have to create `connectionPool` here, right? Its already created as part of `PostgresClient` and its accessed via getPooledConnection.

Comment (suresh-prakash) [reply document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresDatastore.java]: Good catch! Updated.

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresConnectionStringGetter.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+  private String buildConnectionStringIfRequired() {
+    return Optional.of(config)
+        .map(PostgresConfig::getUrl)
+        .or(() -> Optional.of(buildConnectionString(config)))
+        .map(url -> appendDatabase(url, config))
+        .orElseThrow();
+  }
+
+  private String buildConnectionString(final PostgresConfig config) {
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresConnectionStringGetter.java:29]:
```suggestion
  private static String buildConnectionString(final PostgresConfig config) {
```

File: document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresConnectionStringGetter.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    return String.format("jdbc:postgresql://%s:%d/", config.getHost(), config.getPort());
+  }
+
+  private String appendDatabase(final String baseConnectionString, final PostgresConfig config) {
```
Comment (kotharironak) [inline document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresConnectionStringGetter.java:33]:
```suggestion
  private static String appendDatabase(final String baseConnectionString, final PostgresConfig config) {
```

**Review by kotharironak — APPROVED** [review]
LGTM, good one.

**Pull request merged**